src/frontend.c

```diff
@@ -82,7 +90,210 @@ char *logfile_path = "out.txt";
+    "By Matthew Beaudouin-Lafon, Evan New-Schmidt, and Adam Novotny\n\n"
+    "A brief introduction to the interface:\n"
+    "- arrow keys to move the cursor\n"
+    "- enter a character by typing it, unless the mode specifies otherwise\n"
+    "- <TAB> to switch between modes\n"
+    "- <CTRL-C> to quit\n"
+    "- <CTRL-R> to read from the file\n"
+    "- <CTRL-S> to write to the file\n"
+    "- <PGUP>/<PGDOWN> move up/down a screen height\n"
+    "- <SHIFT-LEFT>/<SHIFT-RIGHT> move left/right a screen width\n";
+
+// representation of desired state from the cmdline interface
+// see the creatively-named "arguments" struct in `main` for the defaults
+typedef struct {
+  int height, width;    // dimensions of initial canvas (use `DEFAULT_` if 0)
+  int x, y;             // initial cursor position
+  bool load_file;       // read from file at start
+  bool use_stdin;       // read from stdin instead of file_name
+  char *filename;       // file to save to/read from
+  bool connect_remote;  // connect to a remote server
+  char *remote_host;
+  char *remote_port;
+} arguments_t;
+
+/* Make sure the arguments struct is valid and in a non-conflicting state.
+ *
+ * Return values are meant for an argp parser function.
+ *
+ * TODO: fill this out
+ */
+int validate_arguments(arguments_t *arguments) { return 0; }
+
+void parse_args(int argc, char *argv[], arguments_t *arguments) {
+  struct arg_lit *help, *version, *usage;
+  struct arg_int *width, *height;
+  struct arg_file *file;
+  struct arg_str *server, *port;
+  struct arg_end *end;
+
+  void *argtable[] = {
+      help = arg_litn(NULL, "help", 0, 1, "display this help and exit"),
+      usage =
+          arg_litn(NULL, "usage", 0, 1, "display brief usage info and exit"),
+      version =
+          arg_litn(NULL, "version", 0, 1, "display version info and exit"),
+      width = arg_intn("w", "width", "<n>", 0, 1, "initial width of canvas"),
+      height = arg_intn("h", "height", "<n>", 0, 1, "initial height of canvas"),
+      server = arg_strn("s", "server", "<SERVER>", 0, 1,
+                        "address of server to connect to"),
+      port = arg_strn("p", "port", "<PORT>", 0, 1,
+                      "port of server to connect to (default 5000)"),
+      file = arg_filen(NULL, NULL, "[FILE]", 0, 1,
+                       "filepath for read/write ('-' to read from stdin)"),
+      end = arg_end(20),
+  };
+
+  int nerrors = arg_parse(argc, argv, argtable);
+
+  if (help->count > 0) {
+    printf("Usage: %s [OPTION...] [FILE]\n", program_name);
+    printf("%s -- %s\n\n", program_name, program_description);
+    arg_print_glossary(stdout, argtable, "  %-25s %s\n");
+    printf("\n%s\n", program_doc);
+    printf("Report bugs to %s.\n", program_bug_address);
+    exit(0);
+  }
+
+  if (usage->count > 0) {
+    printf("Usage: %s", program_name);
+    arg_print_syntax(stdout, argtable, "\n");
+    exit(0);
+  }
+
+  if (version->count > 0) {
```

Comment:
running `./collascii --version --usage` only prints the usage due to the `exit(0)` in the usage block. I don't think it's a problem, just checking that it's intended behavior.

Reply:
The behavior I've seen from other programs is `--usage`, `--version`, and `--help` are all mutually exclusive. I tried both `--version` and `--usage` on `man` and collascii built with `argp`, and both of them returned the results for only the first one invoked - I don't know of a way to do that here though, I don't think `argtable` can return their positions.

One option would be to return an error if both are invoked.

Reply:
Alright, sounds like not worth the effort then. Let's stick to current behavior.